main.go

```diff
@@ -169,36 +373,115 @@ func NewInternodeDialContext(dialTimeout time.Duration) DialContext {
 
 // Reader to read random data.
 type netperfReader struct {
-	doneCh <-chan struct{}
-	buf    []byte
+	buf []byte
+
+	Host   string
+	Addr   string
+	count  atomic.Uint64
+	TX     atomic.Uint64
+	TTFBH  atomic.Int64
+	TTFBL  atomic.Int64
+	Errors atomic.Uint64
+}
+
+type AsyncReader struct {
+	pr             *netperfReader
+	i              int64 // current reading index
+	prevRune       int   // index of previous rune; or < 0
+	TTFBRegistered bool
+	Start          time.Time
+	ctx            context.Context
 }
 
-func (m *netperfReader) Read(b []byte) (int, error) {
-	select {
-	case <-m.doneCh:
-		return 0, io.EOF
-	default:
+func (a *AsyncReader) Read(b []byte) (n int, err error) {
+	if !a.TTFBRegistered {
+		a.TTFBRegistered = true
+		since := time.Since(a.Start).Milliseconds()
+		if since > a.pr.TTFBH.Load() {
+			a.pr.TTFBH.Store(since)
+		}
+		if since < a.pr.TTFBL.Load() {
+			a.pr.TTFBL.Store(since)
+		}
+	}
+
+	if !stream {
+		if a.i >= int64(len(a.pr.buf)) {
+			return 0, io.EOF
+		}
+		a.prevRune = -1
+		n = copy(b, a.pr.buf[a.i:])
+		a.i += int64(n)
+	} else {
+		if a.ctx.Err() != nil {
+			return 0, io.EOF
+		}
+		n = copy(b, a.pr.buf)
 	}
-	n := copy(b, m.buf)
-	atomic.AddUint64(&dataOut, uint64(n))
+
+	a.pr.TX.Add(uint64(n))
 	return n, nil
 }
 
-func runClient(host string) {
-	host = net.JoinHostPort(host, port)
-	proc := 32
+func NewNetPerfReader(host string) (r *netperfReader) {
+	r = new(netperfReader)
+	r.Host = host
+	r.buf = make([]byte, payloadMB*humanize.MiByte)
+	r.TTFBH.Store(0)
+	r.TTFBL.Store(999999)
```

Comment:
What is this? A deliberate number, or a specific duration? Specify with time.Duration to make it clearer. I guess it is 999 seconds?

Reply:
it's just a max that is way above that we would ever see as the max, it's just an easy way to initialize a "lowest" default value.

Reply:
Yeah. My point was more to make it clear in the code what the number is. `math.MaxInt64` (or whatever matches the type) is self-explanatory. This is "just a number".